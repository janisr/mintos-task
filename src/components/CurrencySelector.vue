<template>
  <section>
    <currency-list
      :currencies="selectedCurrencies"
      :type="'selected-currencies'"
      @currency-selected="toggleCurrency($event)"
    ></currency-list>

    <div class="seperator"></div>

    <currency-list
      :currencies="allCurrencies"
      :type="'all-currencies'"
      @currency-selected="toggleCurrency($event)"
    ></currency-list>
  </section>
</template>

<script lang="ts">
import { Options, Vue } from "vue-class-component";
import CurrencyList from "./CurrencyList.vue";

import { CurrencyType } from "./currency.types";
import CurrencyItem from "./CurrencyItem.vue";

@Options({
  components: { CurrencyList, CurrencyItem },
  inject: { currencies: { default: [] } }
})
export default class CurrencySelector extends Vue {
  public currencies!: CurrencyType[];

  public allCurrencies: CurrencyType[] = [];

  public get selectedCurrencies(): CurrencyType[] {
    return this.allCurrencies?.filter(c => c.selected);
  }

  public toggleCurrency(label: string) {
    this.allCurrencies = this.allCurrencies.map(c => {
      if (c.label === label) {
        c.selected = !c.selected;
      }

      return c;
    });
  }

  beforeMount() {
    this.allCurrencies = [...this.currencies];
  }
}
</script>

<style lang="scss" scoped>
section {
  width: 360px;
  border: 1px solid #969696;
  border-radius: 4px;
  box-shadow: 2px 2px 7px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  padding: 8px;
}

.seperator {
  height: 8px;
}
</style>
