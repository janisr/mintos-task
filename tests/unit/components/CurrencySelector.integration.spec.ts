import { CurrencyType } from "@/components/currency.types";
import CurrencySelector from "@/components/CurrencySelector.vue";
import { mount } from "@vue/test-utils";

describe("CurrencySelector.vue", () => {
  it("should render provided currencies", () => {
    const currencies: CurrencyType[] = [{ label: "EUR", selected: false }];

    const wrapper = mount(CurrencySelector, {
      data: () => {
        return { currencies };
      },
    });

    expect(wrapper.vm.currencies).toEqual(currencies);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render selected currencies", () => {
    const currencies: CurrencyType[] = [{ label: "EUR", selected: true }];

    const wrapper = mount(CurrencySelector, {
      data: () => {
        return { currencies };
      },
    });

    expect(wrapper.vm.allCurrencies).toEqual(currencies);
    expect(wrapper.vm.selectedCurrencies).toEqual(currencies);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should add selected currency", async () => {
    const currencies: CurrencyType[] = [
      { label: "EUR", selected: false },
      { label: "SEK", selected: false },
    ];

    const wrapper = mount(CurrencySelector, {
      data: () => {
        return { currencies };
      },
    });

    const expected: CurrencyType[] = [{ label: "EUR", selected: true }];

    expect(wrapper.vm.allCurrencies).toEqual(currencies);
    expect(wrapper.vm.selectedCurrencies).toEqual([]);

    wrapper.find("ul li.all-currencies:first-child").trigger("click");

    expect(wrapper.vm.selectedCurrencies).toEqual(expected);
  });

  it("should remove selected currency", async () => {
    const currencies: CurrencyType[] = [
      { label: "EUR", selected: true },
      { label: "SEK", selected: false },
    ];

    const wrapper = mount(CurrencySelector, {
      data: () => {
        return { currencies };
      },
    });

    const expected: CurrencyType[] = [{ label: "EUR", selected: true }];

    expect(wrapper.vm.allCurrencies).toEqual(currencies);
    expect(wrapper.vm.selectedCurrencies).toEqual([
      { label: "EUR", selected: true },
    ]);

    wrapper.find("ul li.selected-currencies:first-child").trigger("click");

    expect(wrapper.vm.selectedCurrencies).toEqual([]);
  });
});
