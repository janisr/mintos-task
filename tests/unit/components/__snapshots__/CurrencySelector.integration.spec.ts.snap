// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector.vue should render provided currencies 1`] = `
<section>
  <ul>
    
    
  </ul>
  <div
    class="seperator"
  />
  <ul>
    
    <li
      class="all-currencies"
    >
      <label>
        EUR
      </label>
      <!--v-if-->
    </li>
    
  </ul>
</section>
`;

exports[`CurrencySelector.vue should render selected currencies 1`] = `
<section>
  <ul>
    
    <li
      class="selected selected-currencies"
    >
      <label>
        EUR
      </label>
      <span
        class="remove"
      >
        X
      </span>
    </li>
    
  </ul>
  <div
    class="seperator"
  />
  <ul>
    
    <li
      class="selected all-currencies"
    >
      <label>
        EUR
      </label>
      <!--v-if-->
    </li>
    
  </ul>
</section>
`;
